### PR TITLE
[2.2.x] Virtual notebook

### PR DESCRIPTION
**Pull request by echarles**
## References

Fixes https://github.com/jupyterlab/jupyterlab/issues/8971

- [x] Intersection Observer
- [x] Placeholder
- [x] Loading indicator
- [x] User settings for `CELL_NUM_DIRECT_RENDER`, `RENDER_CELLS_ON_IDLE`, `NON_OBSERVERED_BOTTOM_MARGIN`

## Code changes

This uses an intersection observer 

## User-facing changes

- Placeholder for non rendered cells
- `Loading` indicator if notebook not fully rendered.

## Backwards-incompatible changes

None.

### CODE REVIEW

**Comment by echarles** [comment]
@isabela-pf This PR is about showing content to the user as soon as possible, while the rest of the notebook is being rendered by the browser if it has time for. Once the initial display is done, the user is free to use the notebook to execute cells... 

I'd love your intake on the way we can inform the user:

- That the notebook is not fully loaded.
- For the search functionality, we are thinking to disable it until full completion, or clearly indicate that the search is partial would not. How can we better tackle those options?

**Comment by goanpeca** [comment]
@echarles could you post an image of the placeholders or maybe a small Gif (I have used this https://www.cockos.com/licecap/) to show how this PR looks would be really helpful :-)

**Comment by echarles** [comment]
This is what it is for now. The user is not supposed to see that unless he scrolls very fast.

Screenshot 2020-09-09 at 18 43 55

**Comment by echarles** [comment]
Hi @isabela-pf, The following gif tries to showcase a typical behavior of a notebook loading. Normally the user would not see the placeholder, but if he scrolls down very quickly just after opening the notebook, he could... I have also put a random icon at a random place supposed to indicate the user that the notebook is still loading with a hover. Not sure if hover is good enough. Not sure if the indicator should disappear or change after notebook is loaded.

ezgif-4-73ce736be77c

**Comment by isabela-pf** [comment]
Hi, all! I think notifying the user of a something like this is a great use of the toast notifications (#698) in JupyterLab. Since toasts are made to automatically disappear after a short period of time (I think it’s 15 seconds), and notebooks that need to take advantage of the feature are large enough it will likely take far more time to load, we need to use toasts thoughtfully to ensure they give appropriate feedback throughout the process.

The first toast the user will see should state essentially what you told me: “Due to its size, this notebook is still loading. All searches will be partial until loading completes.”

Toast 1

Once loading is completed, I think there should be another toast that says “This notebook has loaded successfully. All searches will now be complete.”

Toast 2

The only problem I could see with this approach is that these short-lived toasts seem more likely to be missed when they are bookending a process that could take a long time. Think of a user who sees the first toast, but misses the second. They might think their notebook never loaded successfully or that their search results still weren’t comprehensive. Does this seem like a probable issue? 

I’m trying to keep the language as short and direct as possible, but I’d love feedback on whether or not this makes sense.

**Comment by echarles** [comment]
Thx @isabela-pf for this. Maybe I should have said that everytime the notebook (even if quite small) is larger a predefined number of cells, the rendering of the remaining cells will be done in the background. So we need to take care that the warning is not shown on nearly all the notebooks. This implies also that a timeout of eg. 1 second should be introduced (show the indication only after 1 second).  Another question is the availability of toasts https://github.com/jupyterlab/jupyterlab/issues/689

**Comment by isabela-pf** [comment]
That's great feedback! You're right about toasts' current status in JupyterLab. That's my mistake, I thought they were finally implemented a month ago for some reason. Thinking of alternate solutions now.

**Comment by isabela-pf** [comment]
It doesn’t seem like users need to know the notebook is sort of lazy loading until their experience is impacted, so maybe the solution is there. There’s still the problem of current JupyterLab UI not having anything that is suited for this type of temporary status, so this is unfortunately a kind of one-off solution. 

The next thing that makes the most sense to me is a message that appears in response to a search initiated while a notebook isn’t fully loaded. This way feedback is provided only when it is relevant/impacting users, not all the time (since I mistook which notebooks are taking advantage of this treatment with my last approach).

I mocked this up in the Find interface (all of which could use some polishing, I promise it’s on my to do list) since that was the first place that came to mind for an in-notebook search. Mockup is using `--jp-ui-font-size0` as we sometimes do for small messages like this. Any of the other Find/Replace content should take it's usual place just below the input box and above the message as it currently does if expanded.

Search Message 5

Please let me know if there are other areas I should be accounting for. Does this seem like it will provide sufficient feedback?

**Comment by echarles** [comment]
I love this proposal. Easy and simple, and as usual, the simplest and easiest solutions are the hardest to find. Thx a lot @isabela-pf for this!

**Comment by mlucool** [comment]
As a note, if a user uses native find, they wouldn't know it was impacting them. I don't think this is a dealbreaker, but just wanted to point it out.

**Comment by isabela-pf** [comment]
Thanks for pointing that out @mlucool! It sounds like this is a good solution, especially to start since this is a new feature, but it's worth thinking about an additional way to indicate whether or not a notebook is fully loaded (this may impact certain extensions, too, and they wouldn't know or might not have time to update with a message). I'll put that on my list.

**Comment by krassowski** [comment]
Interesting work! I am afraid it might break https://github.com/krassowski/jupyterlab-lsp as it is now. Would it be useful to create a signal `cellRendered(cell)` and emit it from `_renderPlaceholderCells`? Sorry if this was proposed already and I missed it!

**Comment by echarles** [comment]
> Interesting work! I am afraid it might break https://github.com/krassowski/jupyterlab-lsp as it is now. Would it be useful to create a signal cellRendered(cell) and emit it from _renderPlaceholderCells? Sorry if this was proposed already and I missed it!

I can add that signal of course. Would it be enough to not break `jupyterlab-lsp`?

**Comment by isabela-pf** [comment]
In addition to the message in the find panel, there’s a couple of other places I can imagine indicate the delayed load. One question first: do initially unloaded cells only load if the user scrolls to where they would be visible, or do they eventually load no matter what the user does?

As for another loading indicator, the the only other approach I see any potential in is having the notebook icon be replaced by a loading indicator until all content has loaded. 

Tab Indicator 1

I can imagine a few downsides to this approach, but I’m proposing it to keep thoughts on this moving. For large notebooks that could take more time to load, this might be a confusing switch from the known experience if they see this icon for a while. 

Also, don’t let this work block the PR. I think starting with the message in find where we know experience will be changed is good work starting out.

**Comment by echarles** [comment]
> One question first: do initially unloaded cells only load if the user scrolls to where they would be visible, or do they eventually load no matter what the user does?

They will be loaded no matter what the user does.

>  having the notebook icon be replaced by a loading indicator until all content has loaded.

That is a top proposal. Once again, I am sure it took you ages to converge to this. The simplest and nicest solutions are the ones that take the most time to design.

> I can imagine a few downsides to this approach, but I’m proposing it to keep thoughts on this moving. For large notebooks that could take more time to load, this might be a confusing switch from the known experience if they see this icon for a while.

True. Especially because the user will see on his screen the first cells anyway, and also that loading icon. Maybe a hoover on that icon to show `This notebook is not fully loaded yet`. On the other hand, it is not that bad when you think the browser experience: this has made the success of the Internet: The page is still loading, but you can already read the content. For former slow network, this was a clear winner that has allowed to get the Browser on every desktop.

Another potential downside is that that icon would be flipping for most of the cases. It would show up for eg 0.1 second, and  then be replaced by the classic icon. I can mitigate that by showing the loading icon for notebook that take more than 1 second to load. But then the question is which icon to display between 0 and 1 second?

> Also, don’t let this work block the PR. I think starting with the message in find where we know experience will be changed is good work starting out.

Sure, but great that you continue posting here with any idea/proposal. I will see if I can pull that feature in this PR. Otherwise I will open an issue to keep track of it and that could be done in another PR.

**Comment by echarles** [comment]
@krassowski I have tried the lsp extension and it seems to work with the current implementation of virtual-notebook without any signal

Screenshot 2020-09-22 at 18 44 54

**Comment by krassowski** [comment]
The screenshot does not appear to represent the jupyterlab-lsp completer - this is all the JupyterLab built-in completer with completions coming from the kernel. The LSP completer on 2.2 looks like this:

image

(potentially with a different icons theme by default)

**Comment by echarles** [comment]
@krassowski I was using `@krassowski/jupyterlab-lsp` version `1.1.2`. Have upgraded to version `2.0.7` and now it gives the expected completion in virtual mode (without any change).

Screenshot 2020-09-23 at 09 16 35

**Comment by echarles** [comment]
cc/ @afshin

**Review by saulshanabrook — DISMISSED** [review]
This looks good overall, thank you Eric!

I left a few small style comments, but nothing show stopping.

However, I was testing it out locally and seemed to find one case of odd behavior. If I copy and paste a large group of cells repeatedly, they end up never being rendered? [Here](https://drive.google.com/file/d/1_jzSvCOpdcJ7cACPz9UmGRw6urkqrCez/view?usp=sharing) is a GIF showing it, you can see if I only copy a few it works out OK. But once I do a bunch it seems to then get stuck somehow.

We can postpone this for a follow up, but having some indicator of the total number of cells vs the numbered rendered might be useful, at least for debugging :)

**Comment by echarles** [comment]
> However, I was testing it out locally and seemed to find one case of odd behavior. If I copy and paste a large group of cells repeatedly, they end up never being rendered? Here is a GIF showing it, you can see if I only copy a few it works out OK. But once I do a bunch it seems to then get stuck somehow.

Thx for pushing the boundaries. Probably cells being pasted should be rendered anyway, this should solve the issue. I need to debug a bit more (my tomorrow) to find other ideas.

Will also bring changes to address your comments. Thx again!

> We can postpone this for a follow up, but having some indicator of the total number of cells vs the numbered rendered might be useful, at least for debugging :)

Are you meaning the total number of cells visible vs the numbered rendered to be displayed in e.g. the toolbar for the user, or just as in the browser console?

**Comment by saulshanabrook** [comment]
@echarles 

> Are you meaning the total number of cells visible vs the numbered rendered to be displayed in e.g. the toolbar for the user, or just as in the browser console?

I meant for the user, but no worries about it for now, might not make much sense for normal users would just be helpful when I was trying to debug 🤷

File: packages/notebook-extension/schema/tracker.json
Comment (echarles) [reply packages/notebook-extension/schema/tracker.json:380]: This is more for power user that would have an understanding of how the virtual notebook works and want to optimize the rendering. e.g. you could set 1000px and the observed area would be 1000px down the bottom off the displayed notebook. All cells being added in that area would be observed and rendered. This is really a technical parameter. I am going to add more description for this params so that they will be readable in the user setting page.

File: packages/notebook/src/widget.ts
Comment (echarles) [reply packages/notebook/src/widget.ts:45]: Removed. thx.

File: packages/notebook/src/widget.ts
Comment (echarles) [reply packages/notebook/src/widget.ts:194]: Great call! I have implemented your both suggestions.

File: packages/notebook/src/widget.ts
Comment (echarles) [reply packages/notebook/src/widget.ts:906]: Done. Have also put a bit more details to better explain those settings. Thx!

File: tsconfigdoc.json
Comment (echarles) [reply tsconfigdoc.json:66]: Looks like jlpm has done it for me. I will revert it back as this is not related to this PR. Thx!

**Comment by echarles** [comment]
@saulshanabrook I have implemented your suggestions, added a section in the docs, and fixed the copy/paste issues your have highlighted. I have also fixed a weird behavior when you were inserting a cell (eg. with 'B' keystroke), a place holder was inserted and directly rendered, but this was not the best user experience.

To fix the `copy/paste` and the `insert`, I had to identify in a way what type of actions was requested. The notebook has no knowledge of the document itself, it just responds to commands like push, insert... The ObservableList does not make any distinction on the emitted signal (both `push` and `insert` result in a `add` type signal). I didn't want to add/change those signal types, so I have used a artifice to set the `oldIndex` to `-1` for push(All) and `-2` for insert(All). I have added a note in the code for this and updated the test.

**Comment by echarles** [comment]
The `js-notebook` failures are not related to this PR, see also https://github.com/jupyterlab/jupyterlab/pull/8968#issuecomment-703082121

**Comment by echarles** [comment]
I have opened https://github.com/jupyterlab/jupyterlab/issues/9136 to track the CI build failures which are I believe unrelated to this PR. To me this PR is ready for final review.

**Comment by echarles** [comment]
Updated the status regarding the CI failures on https://github.com/jupyterlab/jupyterlab/issues/9136#issuecomment-703216288

The `js-notebook` and `docs` failure shown here are not related to this PR and should not block the merge of it.

**Comment by afshin** [comment]
I compared Chrome and Firefox on macOS and the behavior on Chrome stutters a bit with a scrollbar behavior that is not what I expected. But the content does render and show. I'm not sure if it's a problem:

scroll-chrome

Comparatively, Firefox's scrolling works exactly as I expect:

scroll-firefox

**Comment by echarles** [comment]
Thx for testing and pointing this @afshin. The stuttering behavior of the scrollbar is due to the `content-visibility` new css property introduced in https://github.com/jupyterlab/jupyterlab/pull/8970

If you want to try out this PR without the. content-visibility, just remove line 129 of `notebook/style/base.css`. You will see that chrome scrollbar just behaves like before, which is good as the virtual notebook does not introduce any change.

As content-visibility is not supported by firefox, you have not seen that in firefox

I am happy to temporary rollback the content-visibility attribute so we can take time to see if we have a fix for this, or discuss if we should or should not accept that change. To start the discussion, I can already say that this virtual notebook PR implements a similar/better logic as the content-visibility.

We have a 2.3 release session in a few hours, so it would be great to have a decision soon to temporary rollback content-visibility or not.

**Comment by echarles** [comment]
Just opened https://github.com/jupyterlab/jupyterlab/pull/9139 to remove content-visibility if we think we should (independent of this PR).

**Comment by echarles** [comment]
@afshin I have rebased on top of https://github.com/jupyterlab/jupyterlab/pull/9139

Any remaining question before we get this merged?

File: packages/notebook/src/widget.ts
Comment (afshin) [inline packages/notebook/src/widget.ts:146]: Please update this

**Review by afshin — APPROVED** [review]
I left one comment re: docstring. Otherwise, 👍 
Thank you for this! 🔥

File: packages/notebook/src/widget.ts
Comment (echarles) [reply packages/notebook/src/widget.ts]: thx. Pushed a docstring.

**Comment by echarles** [comment]
Thx a lot @afshin :heart:

**Comment by echarles** [comment]
For a large notebook that would take e.g. 1 minute with the waiting rolling icon and no cell displayed, you could go to 10 second to view the beginning of the notebook (not fully rendered, only the e.g. 7 cells) but usable and all along the rest of the 50 minutes more and more cells being rendered (until they are all).

**Pull request merged**
